### PR TITLE
Run device comparison test in tape mode

### PR DESCRIPTION
**Pull request by trbromley**
We are currently seeing the `cirq.simulator` device failing with a [strange error](https://github.com/PennyLaneAI/plugin-test-matrix/runs/1665972116?check_suite_focus=true) when tested against the shared device tests. The `test_four_qubit_random_circuit` test in particular is the one failing.

This test works by randomly generating parameters on the [line](https://github.com/PennyLaneAI/pennylane/blob/f4f2bd4bec288948118962b6d7227b3d16e88b0b/pennylane/devices/tests/test_compare_default_qubit.py#L191) `params = list(np.pi * np.random.rand(gate.num_params))`, which is placed *within* the QNode function (`circuit()`). Here, `params` is a list of PennyLane tensors.

In non-tape mode, parameters are expected to be initialized outside of the QNode and are then unwrapped to base NumPy arrays when sent to the device. By defining the parameters as we do in the test above, we see them reach the device in PennyLane Tensor form. In the Cirq device, most gates do not seem to mind working with a PennyLane tensor, but at least the `rz` gate finds this representation problematic and causes an error.

To fix this, we could either:
1. Have the test convert to a list of NumPy arrays
2. Add a line in the Device to ensure we convert to a list of NumPy arrays
3. Perform the test in tape mode

Indeed, this issue is strongly linked to https://github.com/PennyLaneAI/pennylane/pull/941, which ensures that unwrapping always occurs in tape mode.

Since we are anyway moving towards tape mode, I propose to adopt option 3 in this PR. Keeping the defined parameters as a list of PennyLane tensors is anyway a nice compatibility test.

### CODE REVIEW

File: pennylane/devices/tests/test_compare_default_qubit.py [hunk 1]
```diff
@@ -145,57 +145,68 @@ def circuit(weights):
...
-            pytest.skip("Device is default.qubit.")
 
-        if not dev.analytic:
-            pytest.skip("Device is in non-analytical mode.")
+        tape_mode_active = qml.tape_mode_active()
+
+        if not tape_mode_active:
+            qml.enable_tape()
+
+        try:
```
Comment (trbromley) [inline pennylane/devices/tests/test_compare_default_qubit.py:154]: This is the new stuff, everything within the `try` is just an indent of before.

File: pennylane/devices/tests/test_compare_default_qubit.py [hunk 2]
```diff
@@ -145,57 +145,68 @@ def circuit(weights):
...
+                return qml.expval(qml.PauliZ(0))
+
+            qnode_def = qml.QNode(circuit, dev_def)
+            qnode = qml.QNode(circuit, dev)
+
+            assert np.allclose(qnode(), qnode_def(), atol=tol(dev.analytic))
+
+        finally:
+            if not tape_mode_active:
+                qml.disable_tape()
```
Comment (trbromley) [inline pennylane/devices/tests/test_compare_default_qubit.py:212]: This is also new.

**Comment by trbromley** [comment]
Note, I'm not sure if this PR alone will fix the [Cirq](https://github.com/PennyLaneAI/plugin-test-matrix#testing-matrix) row :thinking:
I was seeing some issues with the `qsim` device:
`pl-device-test --device=cirq.qsim --tb=short --skip-ops --analytic=True`
image

I believe these are unrelated issues though.

**Review by josh146 — APPROVED** [review]
Thanks for fixing this @trbromley!

I have a very minor suggestion, which is to simply add a new fixture

```python
@pytest.fixture()
def tape_mode_only(self):
    qml.enable_tape()
    yield
    qml.disable_tape()
```

to avoid the indentation and try-except. But this is minor, and not a barrier to merging.

**Review by antalszava — APPROVED** [review]

File: pennylane/devices/tests/test_compare_default_qubit.py [hunk 3]
```diff
@@ -145,57 +145,70 @@ def circuit(weights):
...
-        qnode_def = qml.QNode(circuit, dev_def)
-        qnode = qml.QNode(circuit, dev)
-
-        assert np.allclose(qnode(), qnode_def(), atol=tol(dev.analytic))
+            gates_per_layers = [np.random.permutation(gates).numpy() for _ in range(layers)]
+
+            def circuit():
+                """4-qubit circuit with layers of randomly selected gates and random connections for
+                multi-qubit gates."""
+                np.random.seed(1967)
```
Comment (antalszava) [inline pennylane/devices/tests/test_compare_default_qubit.py:195]: (Minor:) Not entirely in scope (as this was not changed), but do we need this seeding and the one on line 189 too? 🤔 
```suggestion
```

**Comment by josh146** [comment]
@trbromley it looks like there are merge conflicts --- I'm not entirely sure the best way to resolve them

**Comment by trbromley** [comment]
Closing this PR: the `cirq` tests are now all passing, and I don't see this problem present in other tests.

**Pull request closed without merging**